db/migrate/20170221195674_tidy_current_roles.rb

```diff
@@ -0,0 +1,45 @@
+class TidyCurrentRoles < ActiveRecord::Migration
+  def up
+    Role.without_auditing do
+      Foreman::Constants::RolesList.roles.each do |name, permission_names|
+        role = Role.find_by :name => name
+        if role
+          process_existing name, role, permission_names
+        else
+          create_from_seeds name, permission_names
+        end
+      end
+    end
+  end
+
+  def process_existing(original_name, role, permission_names)
+    diff = role.permission_diff permission_names
+    return if diff.empty?
+    rename_existing role, original_name
+    create_from_seeds original_name, permission_names
+  end
+
+  def rename_existing(role, original_name)
+    prefix = "Custom"
+    role.name = "#{prefix} #{original_name}"
+    if role.valid?
+      role.save!
+    else
+      rename_with_valid_name role, prefix, original_name
+    end
+  end
+
+  def rename_with_valid_name(role, prefix, original_name)
+    num = 0
+    until role.valid?
```

Comment:
What if role is invalid not because of the name? this would never exit the loop. I think we the migration should not even try to do the validations. The reason is that the migration can become broken later when Role model would introduce new validations. I'd recommend changing the code so that it find the first name that can be used (finding using active recrod is safe). Then do the role update skipping validations and callback, normally we use `update_all`, e.g. `Role.where(:id => role.id).update_all(:name => found_free_name)`